### PR TITLE
[sBTC DR] Add proof data for verification in clarity, add verification in mint and burn function

### LINKED ISSUE #79: [sBTC DR] Assert deposit/witdhdrawal tx exists in mint/burn contract calls
### Definition of done
The sBTC DR asset contract does not allow mints and burns that do not refer to existing bitcoin transactions

### PR DESCRIPTION
**Pull request by friedger**
This PR
* adds type ProofData that can be used for verification of bitcoin txs in Clarity
* adds tx verification for mint and burn
* fixes #79

### CODE REVIEW

**Review by stjepangolemac — COMMENTED** [review]
Looks good! Two things to address:

- Some of the changes done here are already done in https://github.com/stacks-network/sbtc/pull/92 - branch of that PR is probably a better base for this one than master
- Ignored tests inside romeo were used just to trigger some functions and manually inspect how they're working, they will probably be removed at some point but it's okay to tidy them up for now

**Review by stjepangolemac — DISMISSED** [review]

File: romeo/src/proof.rs [hunk 1]
```diff
@@ -0,0 +1,52 @@
+//! Proof
+use bdk::bitcoin::{Block, BlockHash, Transaction, Txid as BitcoinTxId};
+
+/// A proof for a bitcoin transaction used by clarity contracts
+#[derive(serde::Serialize, serde::Deserialize, Debug, Clone)]
+pub struct Proof {
```
[the remark this answers is not in the document]
Comment (friedger) [reply romeo/src/proof.rs:6]: `Proof` is just a data structure that is send to the clarity engine that will verify whether the proof is valid.

[the remark this answers is not in the document]
Comment (friedger) [reply romeo/src/proof.rs:6]: Should I rename it to `ProofData` ?

[the remark this answers is not in the document]
Comment (friedger) [reply romeo/src/proof.rs]: Renamed!

File: romeo/src/proof.rs [hunk 2]
```diff
@@ -0,0 +1,52 @@
...
+
+// test module
+#[cfg(test)]
+// test from_block returns correct Proof
+mod tests {
+    use super::*;
+    use bdk::bitcoin::{consensus::deserialize, hashes::hex::FromHex, Block};
+
+    #[test]
+    fn should_create_correct_proof() {
```
[the remark this answers is not in the document]
Comment (friedger) [reply romeo/src/proof.rs]: added more tests

File: romeo/src/system.rs [hunk 3]
```diff
@@ -182,7 +185,23 @@ async fn deploy_asset_contract(config: &Config, client: LockedClient) -> Event {
...
+async fn mint_asset(
+    config: &Config,
+    bitcoin_client: BitcoinClient,
+    stacks_client: LockedClient,
+    deposit_info: DepositInfo,
+) -> Event {
+    let block = bitcoin_client
+        .fetch_block(deposit_info.block_height)
+        .await
+        .unwrap();
```
[the remark this answers is not in the document]
Comment (friedger) [reply romeo/src/system.rs]: replaced with expect

File: romeo/src/system.rs [hunk 4]
```diff
@@ -182,7 +185,23 @@ async fn deploy_asset_contract(config: &Config, client: LockedClient) -> Event {
...
+    let index = block
+        .txdata
+        .iter()
+        .position(|tx| tx.txid() == deposit_info.txid)
+        .unwrap();
+    let _proof = Proof::from_block_and_index(&block, index);
```
[the remark this answers is not in the document]
Comment (friedger) [reply romeo/src/system.rs]: renamed to proof_data

File: romeo/src/proof.rs [hunk 5]
```diff
@@ -0,0 +1,52 @@
...
+    pub txid: BitcoinTxId,
+    /// The block height of the bitcoin transaction
+    pub block_height: u64,
+    /// The block hash of the bitcoin transaction
+    pub block_hash: BlockHash,
+}
+
+impl Proof {
+    /// Create a new proof from a bitcoin transaction and a block
+    pub fn from_block_and_index(block: &Block, index: usize) -> Self {
```
[the remark this answers is not in the document]
Comment (friedger) [reply romeo/src/proof.rs]: renamed to proof data

File: romeo/src/system.rs [hunk 6]
```diff
@@ -182,7 +185,23 @@ async fn deploy_asset_contract(config: &Config, client: LockedClient) -> Event {
     Event::AssetContractCreated(txid)
 }
 
-async fn mint_asset(config: &Config, client: LockedClient, deposit_info: DepositInfo) -> Event {
+async fn mint_asset(
+    config: &Config,
+    bitcoin_client: BitcoinClient,
+    stacks_client: LockedClient,
+    deposit_info: DepositInfo,
+) -> Event {
```
[the remark this answers is not in the document]
Comment (friedger) [reply romeo/src/system.rs:192]: I have opened an issue for that: https://github.com/stacks-network/sbtc/issues/110

**Review by stjepangolemac — DISMISSED** [review]

**Review by stjepangolemac — DISMISSED** [review]

**Review by stjepangolemac — APPROVED** [review]

**Comment by friedger** [comment]
@CAGS295 Does this PR make now more sense?

**Review by MarvinJanssen — APPROVED** [review]

**Comment by MarvinJanssen** [comment]
🎉

**Pull request merged**